### PR TITLE
feat: consolidate Qwen3 ForcedAligner into analysis service

### PR DESCRIPTION
**Pull request by mhuang74**
## Summary

Consolidates the Qwen3 ForcedAligner from a separate Docker microservice (`services/qwen3/`) into the Analysis Service (`services/analysis/`), then deletes `services/qwen3/` entirely. This reduces operational complexity to a single Docker image while adding a new `FORCED_ALIGNMENT` job type, an `audio align-lrc` CLI command, and automatic language detection for forced alignment.

### 1. Qwen3 ForcedAligner merged into Analysis Service as in-process worker

- **New `ForcedAlignerWrapper`** (`workers/forced_aligner.py`): Lazy-loads the Qwen3ForcedAligner model on first job via double-check locking (`asyncio.Lock`). No internal semaphore — concurrency is controlled externally by `JobQueue`. Raises `RuntimeError` on init failure (matches `AudioSeparatorWrapper` pattern). Cleans up GPU memory via `torch.cuda.empty_cache()`.

- **New `FORCED_ALIGNMENT` job type**: Added to `JobType` enum with `ForcedAlignmentJobRequest` and `ForcedAlignmentOptions` models. New `POST /api/v1/jobs/forced-alignment` API endpoint.

- **Job processing** (`workers/queue.py`): `_process_forced_alignment_job()` downloads audio from R2, resolves transcription audio (prefers vocals stem, auto-triggers stem separation if needed), validates duration ≤ 300s, runs alignment, maps segments to LRC lines, and uploads LRC to R2. The global `_local_model_semaphore` is acquired **only around the `align()` call** — not the entire job — preventing deadlock with auto-triggered stem separation child jobs.

- **New utility functions** (`workers/forced_alignment.py`): `normalize_text()` (CJK punctuation/whitespace), `format_timestamp()`, `map_segments_to_lines()` (character-level → line-level mapping), `validate_audio_duration()` (hybrid `soundfile.info()` O(1) for WAV/FLAC with `librosa` fallback).

- **Service-level LRC backup**: `R2Client.copy_object()` backs up existing LRC files before overwrite, protecting all callers (API + CLI).

- **DB migration**: Added `_migrate_forced_alignment_type()` for the new job type.

### 2. Auto-detect language for forced alignment

- **`ForcedAlignmentOptions.language`** now accepts `"auto"`, `"zh"`, or `"en"` (default: `"auto"`). When `"auto"`, the analysis service resolves the language using the existing `resolve_lrc_language()` heuristic from `lrc.py`, which inspects both song title and lyrics text for CJK/Latin script. This reuses battle-tested detection (CJK Extensions A–F) instead of duplicating logic.

- **CLI `--lang` default changed** from `"zh"` to `"auto"` — English songs are now correctly detected without manual `--lang en` override.

### 3. Admin CLI: `audio align-lrc` command

- New `sow-admin audio align-lrc <song_id>` command with options: `--lang` (auto/zh/en), `--force`, `--use-vocals-stem/--no-vocals-stem`, `--stdin` (batch mode), `--wait`.
- Looks up recording and song from DB, submits forced alignment job, polls for completion.
- Removed deprecated `--no-qwen3` flag from the `lrc` command.

### 4. Cleanup: `services/qwen3/` deleted entirely

- Removed the entire separate Qwen3 Docker microservice (Dockerfile, docker-compose, routes, workers, storage, config).
- Removed `services/analysis/src/sow_analysis/services/qwen3_client.py` (HTTP client to separate container).
- Removed `docker/docker-compose.prod.yml` (standalone qwen3 prod config).
- Updated `docker-compose.yml`, `.env.example`, `deploy.sh`, and all documentation to replace `SOW_QWEN3_*` config with `SOW_FORCED_ALIGNER_*`.

### 5. Official LRC Last-Writer-Wins (v3)

- **`{hash_prefix}/lyrics.lrc`** is now the single official LRC object. Both regular LRC jobs and ForcedAlignment jobs overwrite this same object.
- **Backup before overwrite**: Before overwriting `lyrics.lrc`, the current official object is copied to `{hash_prefix}/lyrics.backup.{unix_timestamp_ms}.lrc`. Backup failure is fatal (fails the job). A `skip_backup` force flag allows override for emergencies.
- **ETag-based stale-object protection**: All writers (analysis service jobs, cache hits, and admin CLI) capture the ETag of `lyrics.lrc` at operation start. If the object was modified after the operation started, the write fails with `StaleObjectError`.
- **Stopped creating** new `lyrics.{lang}.forced.lrc` and `lyrics.{lang}.v2.lrc` artifacts. Historical `.forced.lrc` and `.v2.lrc` objects in R2 are left untouched.
- **Backup pruning**: After a successful backup, old backups are pruned to keep at most 5 per `hash_prefix`.
- **Cache behavior**: Generated LRC text is now stored in the LRC cache. On cache hit with cached text, the official `lyrics.lrc` is rewritten with ETag check. Metadata-only legacy cache entries are ignored and regenerated (one-time burst after deployment).

### 6. Tests

- **53 new tests** (34 `test_map_segments_to_lines` + 19 `test_forced_alignment`), all passing.
- Coverage includes: model validation, timestamp formatting, duration validation, job processing (success, language mapping, auto-detection, invalid request, alignment failure, backup, deadlock prevention), `ForcedAlignerWrapper` lifecycle.
- Added `upload_official_lrc` tests for both async (analysis) and sync (admin) R2 clients: backup creation, stale ETag detection, backup failure handling, skip-backup flag, pruning.
- Updated queue tests for official LRC upload, cache text rewrite, stale object failure, metadata-only cache ignore.
- Updated LRC worker cache test for text-included cache entries.
- Updated admin editor test for `upload_official_lrc` mock.
- Deleted legacy tests: `test_qwen3_fallback.py`, `test_qwen3_regression.py`, `test_lrc_benchmark.py`, `test_lrc_integration_qwen3.py`.

## Key Design Decisions

| Decision | Choice | Rationale |
|----------|--------|-----------|
| Model loading | Lazy (on first forced-alignment job) | Matches `AudioSeparatorWrapper` pattern; avoids ~1.2GB RAM at startup |
| Semaphore strategy | Acquire ONLY around `align()` call | Prevents deadlock with nested stem-separation jobs |
| Language detection | Reuse `resolve_lrc_language()` from `lrc.py` | Single source of truth; comprehensive CJK detection (Extensions A–F); uses song title as signal |
| Duration validation | Hybrid: `soundfile.info()` first, `librosa` fallback | O(1) for WAV/FLAC; librosa fallback for MP3 |
| LRC overwrite safety | Service-level copy-before-overwrite | Defense in depth for all callers |
| Official LRC object | Single `lyrics.lrc` per hash_prefix | Simplifies consumer code; all renderers already read this path |
| Stale-object detection | ETag-based (no distributed locking) | Single-instance analysis service makes this sufficient; narrow race window accepted |

## Migration Notes

- Existing jobs with explicit `language="zh"` or `"en"` continue to work unchanged
- Default language changed from `"zh"` to `"auto"` — English songs now auto-detected correctly
- `SOW_QWEN3_BASE_URL` / `SOW_QWEN3_API_KEY` replaced by `SOW_FORCED_ALIGNER_MODEL_PATH` / `SOW_FORCED_ALIGNER_DEVICE`
- Single Docker container (`services/analysis`) now handles all job types including forced alignment
- Historical `.forced.lrc` and `.v2.lrc` objects in R2 are left untouched; no migration needed
- One-time burst of LRC reprocessing will occur after deployment for songs with metadata-only cache entries

## Verification (requires running services)

- [ ] `docker compose up analysis` works without qwen3 service
- [ ] `sow-admin audio align-lrc <song_id> --wait` works end-to-end
- [ ] `sow-admin audio lrc <song_id>` still works unchanged
- [ ] `sow-admin audio upload-lrc <song_id> <file.lrc>` creates backup and respects ETag

### CODE REVIEW

File: services/analysis/src/sow_analysis/workers/forced_alignment.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    aligned_text = ""
+    segment_positions = []
+
+    for seg_start, seg_end, seg_text in segments:
+        start_char = len(aligned_text)
+        aligned_text += seg_text
+        end_char = len(aligned_text)
+        segment_positions.append((start_char, end_char, seg_start, seg_end, seg_text))
+
+    aligned_normalized = normalize_text(aligned_text)
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply services/analysis/src/sow_analysis/workers/forced_alignment.py]: Fixed in bb0a73f. Build aligned_normalized incrementally from normalized segments so segment_positions indices are in normalized space, matching the line_start/line_end lookups.

File: services/analysis/src/sow_analysis/workers/forced_alignment.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    """Validate audio duration using soundfile (O(1) for WAV/FLAC) with librosa fallback."""
+    try:
+        import soundfile
+
+        info = soundfile.info(str(audio_path))
+        duration = info.duration
+    except Exception:
+        import librosa
+
+        duration = librosa.get_duration(filename=str(audio_path))
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply services/analysis/src/sow_analysis/workers/forced_alignment.py]: Fixed in bb0a73f. Changed librosa.get_duration(filename=...) to librosa.get_duration(path=...).

File: services/analysis/src/sow_analysis/workers/forced_aligner.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    async def cleanup(self) -> None:
+        """Unload the model and release resources."""
+        self._ready = False
+        self._model = None
+        gc.collect()
+        try:
+            import torch
+
+            if torch.cuda.is_available():
+                torch.cuda.empty_cache()
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply services/analysis/src/sow_analysis/workers/forced_aligner.py:148]: Fixed in bb0a73f. Added MPS cache cleanup with hasattr guard for torch.mps.empty_cache().

File: services/analysis/src/sow_analysis/storage/r2.py [hunk 4]
```diff
@@ -264,3 +264,22 @@ async def upload_clean_stems(
...
+    async def copy_object(self, source_s3_url: str, dest_s3_url: str) -> None:
+        """Copy an object within R2.
+
+        Args:
+            source_s3_url: Source s3://bucket/key URL
+            dest_s3_url: Destination s3://bucket/key URL
+        """
+        src_bucket, src_key = parse_s3_url(source_s3_url)
+        dest_bucket, dest_key = parse_s3_url(dest_s3_url)
+        loop = asyncio.get_event_loop()
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply services/analysis/src/sow_analysis/storage/r2.py]: Fixed in bb0a73f. Changed asyncio.get_event_loop() to asyncio.get_running_loop() in copy_object().

File: services/analysis/src/sow_analysis/workers/queue.py [hunk 5]
```diff
@@ -1026,6 +1072,196 @@ async def _process_lrc_job(self, job: Job) -> None:
...
+            except Exception as e:
+                logger.error(f"Failed to update job {job.id} in database: {e}")
+            return
+
+        language_map = {"zh": "Chinese", "en": "English"}
+        language_mapped = language_map.get(request.options.language, "Chinese")
+
+        try:
+            if not self.r2_client and settings.SOW_R2_ENDPOINT_URL:
+                self.initialize_r2(settings.SOW_R2_BUCKET, settings.SOW_R2_ENDPOINT_URL)
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply services/analysis/src/sow_analysis/workers/queue.py:1262]: Fixed in bb0a73f. Added early RuntimeError if R2 client is not initialized after the initialization attempt.

File: services/analysis/src/sow_analysis/workers/queue.py [hunk 6]
```diff
@@ -1026,6 +1072,196 @@ async def _process_lrc_job(self, job: Job) -> None:
...
+                    try:
+                        existing_lrc_url = f"s3://{self.r2_client.bucket}/{target_key}"
+                        if await self.r2_client.check_exists(existing_lrc_url):
+                            import time as _time
+
+                            backup_key = (
+                                f"{hash_prefix}/lyrics.{request.options.language}"
+                                f".backup.{int(_time.time())}.lrc"
+                            )
+                            await self.r2_client.copy_object(
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply services/analysis/src/sow_analysis/workers/queue.py]: Fixed in bb0a73f. Removed redundant 'import time as _time' and replaced _time.time() with module-level time.time().

File: services/analysis/start-dev.sh [hunk 7]
```diff
@@ -104,8 +105,53 @@ EOF
...
+
+if [[ "$QWEN3_MODEL_FOUND" == false ]]; then
+    echo -e "  ${YELLOW}Missing: Qwen3-ForcedAligner-0.6B${NC}"
+    echo ""
+    echo -e "${YELLOW}Downloading Qwen3 Forced Aligner model from Hugging Face...${NC}"
+    echo "This may take several minutes (~1.2GB)..."
+    echo ""
+
+    cd "$PROJECT_ROOT"
+    uv run --python 3.11 --extra dev python << EOF
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply services/analysis/start-dev.sh]: Fixed in bb0a73f. Changed --extra dev to --extra poc_qwen3_asr which includes huggingface-hub as a dependency.
